// @ct-shipping/maskit/src/maskIt.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var traverse_1 = __importDefault(require("traverse"));
var full_1 = require("klona/full");
var MaskIt = /** @class */ (function () {
    function MaskIt() {
        this.sensitiveKeys = [
            /passw(or)?d/i,
            /^pw$/,
            /^pass$/i,
            /secret/i,
            /token/i,
            /api[-._]?key/i,
        ];
    }
    MaskIt.prototype.isSensitiveKey = function (keyStr) {
        if (keyStr) {
            return this.sensitiveKeys.some(function (regex) { return regex.test(keyStr); });
        }
        return false;
    };
    MaskIt.prototype.maskObject = function (obj) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        var that = this;
        (0, traverse_1.default)(obj).forEach(function redactor() {
            if (that.isSensitiveKey(this.key)) {
                this.update('[MASKED]');
            }
        });
    };
    MaskIt.prototype.mask = function (obj) {
        var copy = (0, full_1.klona)(obj);
        this.maskObject(copy);
        return copy;
    };
    MaskIt.prototype.addSensitiveKeys = function (obj) {
        var _a;
        (_a = this.sensitiveKeys).push.apply(_a, obj);
    };
    return MaskIt;
}());
var maskit = new MaskIt();
exports.default = maskit;
